portfolio/src/main/java/com/google/sps/data/HighestPriorityFirstScheduler.java

```diff
@@ -0,0 +1,139 @@
+      Comparator.comparing(ScheduledTask::getStartTime);
+
+  /**
+   * This method schedules tasks from highest to lowest priority and shortest to longest duration.
+   */
+  public Collection<ScheduledTask> schedule(
+      Collection<CalendarEvent> events,
+      Collection<Task> tasks,
+      Instant workHoursStartTime,
+      Instant workHoursEndTime) {
+    List<CalendarEvent> eventsList = new ArrayList<CalendarEvent>(events);
+    List<Task> tasksList = new ArrayList<Task>(tasks);
+
+    TaskQueue taskQueue = new TaskQueue(tasksList, getSchedulingAlgorithmType());
+
+    CalendarEventsGroup calendarEventsGroup =
+        new CalendarEventsGroup(eventsList, workHoursStartTime, workHoursEndTime);
+
+    // When accessing available TimeRanges, work with currentAvailableTimes
+    // since availableTimes will not be updated after creation while
+    // currentAvailableTimes will be updated everytime the algorithm works
+    // with a new priority value. Updates included modified or deleted TimeRanges.
+    List<TimeRange> availableTimes = calendarEventsGroup.getFreeTimeRanges();
+    TimeRangeGroup availableTimesGroup = new ArrayListTimeRangeGroup(availableTimes);
+    List<TimeRange> currentAvailableTimes = getAvailableTimeRangesList(availableTimesGroup);
+
+    List<ScheduledTask> scheduledTasks = new ArrayList<ScheduledTask>();
+
+    int currentAvailableTimesIndex = 0;
+
+    Instant currentScheduleTime = workHoursStartTime;
+
+    // The algorithm begins with the Task with the highest priority, and lowest duration. It
+    // then iterates through the currentAvailableTimes list trying to find a TimeRange that is
+    // large enough to fit the Task in and schedules it there. Once a Task is scheduled, the
+    // next Task from the PriorityQueue is retrieved until the PriorityQueue is empty and each
+    // Task that can be scheduled has been scheduled. If a Task has the same priority as the
+    // previously scheduled task then the scheduling algorithm tries to schedule the task starting
+    // at the TimeRange that the previous Task was scheduled in. If the Task has a different
+    // priority
+    // then the algorithm starts at the first available TimeRange.
+    while (currentAvailableTimesIndex < currentAvailableTimes.size() && !taskQueue.isEmpty()) {
+      TimeRange currentAvailableTimeRange = currentAvailableTimes.get(currentAvailableTimesIndex);
+      Task task = taskQueue.peek();
+
+      if (currentAvailableTimeRange.start().isAfter(currentScheduleTime)) {
+        currentScheduleTime = currentAvailableTimeRange.start();
+      }
+
+      boolean doesTaskFit =
+          !currentScheduleTime
+              .plusSeconds(task.getDuration().getSeconds())
+              .isAfter(currentAvailableTimeRange.end());
+
+      if (doesTaskFit) {
+        ScheduledTask scheduledTask =
+            new ScheduledTask(
+                task, currentScheduleTime, /* isCompletelyScheduled = */ Optional.of(true));
```

Comment:
Here I think we can change `Optional.of(true)` to `Optional.of(SchedulingCompleteness.COMPLETELY_SCHEDULED)`